save-orchestrator/src/main/kotlin/com/saveourtool/save/orchestrator/controller/HeartbeatController.kt

```diff
@@ -101,122 +93,76 @@ class HeartbeatController(private val agentService: AgentService,
+                        agentService.updateAgentStatusesWithDto(listOf(AgentStatusDto(LocalDateTime.now(), TERMINATED, agentId)))
+                            .thenReturn(TerminateResponse)
+                            .doOnSuccess {
+                                logger.info("Agent id=$agentId will receive ${TerminateResponse::class.simpleName} and should shutdown gracefully")
+                                ensureGracefulShutdown(agentId)
+                            }
+                    } else {
+                        Mono.just(response)
+                    }
+                }
 
-    /**
-     * Consider agent as crashed, if it didn't send heartbeats for some time
-     */
-    fun determineCrashedAgents() {
-        agentsLatestHeartBeatsMap.filter { (currentAgentId, _) ->
-            currentAgentId !in crashedAgentsList
-        }.forEach { (currentAgentId, stateToLatestHeartBeatPair) ->
-            val duration = (Clock.System.now() - stateToLatestHeartBeatPair.second).inWholeMilliseconds
-            logger.debug("Latest heartbeat from $currentAgentId was sent: $duration ms ago")
-            if (duration >= configProperties.agentsHeartBeatTimeoutMillis) {
-                logger.debug("Adding $currentAgentId to list crashed agents")
-                crashedAgentsList.add(currentAgentId)
-            }
-        }
+    private fun handleFinishedAgent(agentId: String, isSavingSuccessful: Boolean): Mono<HeartbeatResponse> = if (isSavingSuccessful) {
+        handleVacantAgent(agentId, isStarting = false)
+    } else {
+        // Agent finished its work, however only part of results were received, other should be marked as failed
+        agentService.markTestExecutionsAsFailed(listOf(agentId), FINISHED)
+            .subscribeOn(agentService.scheduler)
+            .subscribe()
+        Mono.just(WaitResponse)
     }
 
-    /**
-     * Stop crashed agents and mark corresponding test executions as failed with internal error
-     */
-    fun processCrashedAgents() {
-        if (crashedAgentsList.isEmpty()) {
-            return
-        }
-        logger.debug("Stopping crashed agents: $crashedAgentsList")
-
-        val areAgentsStopped = dockerService.stopAgents(crashedAgentsList)
-        if (areAgentsStopped) {
-            agentService.markAgentsAndTestExecutionsCrashed(crashedAgentsList)
-            logger.warn("All agents are crashed, initialize shutdown sequence")
-            if (agentsLatestHeartBeatsMap.keys().toList() == crashedAgentsList.toList()) {
-                initiateShutdownSequence(crashedAgentsList.first(), areAllAgentsCrashed = true)
-            }
-        } else {
-            logger.warn("Crashed agents $crashedAgentsList are not stopped after stop command")
+    private fun handleIllegallyOnlineAgent(agentId: String, state: AgentState) {
+        logger.warn("Agent id=$agentId sent $state status, but should be offline in that case!")
+        if (agentId !in crashedAgents) {
```

Comment:
we can simplify checking -- Set guarantees that collection doesn't contain duplicate